### PR TITLE
Add env var substitution support to file configuration

### PR DESCRIPTION
**Pull request by jack-berg**
Environment variable substitution was one of the key parts of file configuration described in the [original OTEP](https://github.com/open-telemetry/oteps/blob/main/text/0225-configuration.md#environment-variable-substitution).

I've been holding off defining it in the [file configuration spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/file-configuration.md) because I've been trying to define the key [parse / create operations](https://github.com/open-telemetry/opentelemetry-specification/pull/3437), and it makes sense to describe environment variable substitution as a requirement of `parse`.

So while this isn't part of the spec yet, I will be trying to get it added as soon as possible. I think the syntax of environment variables (i.e. `${env:VARIABLE_NAME}`) is the most likely thing the spec lands on since its the syntax the collector uses and it will be hard to argue that an alternative new syntax would trump consistency with the collector.

### CODE REVIEW

File: sdk-extensions/incubator/src/test/java/io/opentelemetry/sdk/extension/incubator/fileconfig/ConfigurationReaderTest.java
Comment (breedx-splk) [inline sdk-extensions/incubator/src/test/java/io/opentelemetry/sdk/extension/incubator/fileconfig/ConfigurationReaderTest.java:423]: These multi-line yamls that have been collapsed into a single one-line string with `\n` in the middle are a bit hard to read. I think a string concat would read slightly better here.

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/ConfigurationReader.java
Comment (breedx-splk) [inline sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/ConfigurationReader.java:28]: As it reads now, `${ env:MY_VAR }` would fail the parse due to the space after/before the curly braces. I think other systems are pretty tolerant of whitespace inside the `${ ... }` variable marker.  Bash isn't, but python and javascript are.

I'm guessing this isn't clarified in the spec yet? I wouldn't hold up this PR for this, but it would be good to give it some consideration. Even `"\\$\\{\\s*env:([a-zA-Z_]+[a-zA-Z0-9_]*)\\s*}"` might be enough?

**Review by breedx-splk — APPROVED** [review]
Yeah this is great. I think users are going to get a lot of mileage out of this.

**Review by mateuszrzeszutek — APPROVED** [review]
👍 

Do you think it'd make sense to add sth like `${sys:otel.traces.exporter}` for system properties as well?

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/ConfigurationReader.java
Comment (jack-berg) [reply sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/ConfigurationReader.java:28]: I think we start strict and relax later if we need to.

I'm planning on opening a spec PR for this - just working out a few things about how the substitution would work with JSON, which various maintainers insisted on supporting with file configuration.

**Comment by jack-berg** [comment]
I think we probably should @mateuszrzeszutek. Opened #5926 to discuss and track.

**Pull request merged**
